### ISSUE TITLE
trakSTAR sampling issue

### ISSUE THREAD

**Issue by awong26** [BLAM-Lab-Projects/experiment-code#3]
Found a bug where the trakSTAR is actually not sampling as desired. It pulls data, but because it does so one sensor at a time you can end up in a situation where some sensors are updated and others are not. I've implemented a work-around in my TMS code, but this has not been uploaded into the TrackBird object yet. 

Also, because it of the interpolation algorithm, asking for samples at a particular sampling rate actually returns samples at 3x that rate, which is quite fast. As a result, it often drops samples because it cannot poll fast enough (i.e., every ~2 ms) to catch each sample. This is true even if there is only a single bird to poll data from - likely because of all the other computations occurring in the drawing function and the state machine.

The ideal fix is to (1) switch to streaming mode, which allows us to set the data-return rate equal to the sampling rate at which all 3 transmitter axes are sampled (i.e., "true" 130 Hz analogous to the FOB), and/or (2) use group-sampling with GetAsynchronousSample() to capture the data from all 4 birds simultaneously so we don't have odd situations were some of the birds are updated and some are not.

Note, the current code may potentially be a problem for the Zurich SMARTS II kinereach setup, especially if they are not saving the Tracker Time (I can't remember if it does so or not, but even if it does it may not be doing so correctly).

**Comment by awong26** [comment]
Uploaded proposed changes to TrackBird.cpp to handle these issues. In particular, it pulls trakSTAR data from all sensors in a single call, avoiding the issue with sampling each sensor at slightly different times. 

The changes also allow for the possibility of sampling using the GetSynchronousRecord() function call, which makes use of the REPORT_RATE parameter. (Note, this is currently commented out, and the GetAsychnronousRecord() function call is currently used). The GetSynchronousRecord() function allows for data to be pulled at the correct rate (i.e., 140 Hz) while still energizing the transmitter axes at 3x that rate (analogous to the workings of the Flock of Birds). It remains to be verified that this avoids the use of the built-in interpolation algorithm and instead returns real data from each axis.

In summary, changes (1) and (2) have been implemented. They have been tested once on the trakSTAR, but has not been not fully verified.

**Comment by aforren1** [comment]
https://github.com/github/gitignore/blob/master/VisualStudio.gitignore

**Comment by awong26** [comment]
As best as I can tell according to the trakSTAR documentation, the REPORT_RATE simply reduces the number of samples that are sent, but otherwise doesn't change how they are collected. So this doesn't seem to address the interpolation issue. In that case, it makes sense to try to collect samples at the highest sampling rate possible.

**Comment by awong26** [comment]
Decided the best solution for now is to set up the option to stream synchronously (when the REPORTRATE in config.h is set higher than 1), otherwise it will stream asynchronously. Streaming asynchronously drops occasional samples (but not a ton of samples), but the time stamp saved into the data file means we can work around the problem. It is still unclear whether you can get similar dropped-sample (or lag behind) problems when streaming synchronously; however, setting the REPORTRATE above 1 should not drop any samples (since the acquisition rate is then above the rate that you can acquire samples asynchronously) so we should not have an issue.

This issue may be revisited in the future, but we will consider it closed for now.
